osidb/mixins.py

```diff
@@ -657,16 +657,47 @@ def alert(
                 uuid.UUID(acl) for acl in generate_acls([settings.PUBLIC_WRITE_GROUP])
             ]
 
-        Alert.objects.create_alert(
-            name=name,
-            object_id=self.uuid,
-            content_type=ContentType.objects.get_for_model(self),
-            description=description,
-            alert_type=alert_type,
-            resolution_steps=resolution_steps,
-            acl_read=acl_read,
-            acl_write=acl_write,
-        ).save()
+        # Alert.objects.create_alert().save() can raise IntegrityError
+        # (duplicate key value violates unique constraint "unique Alert for name and object").
+        #
+        # The database internally already uses a lock to ensure the uniqueness even when
+        # multiple threads each try to create the same row concurrently.
+        #
+        # Therefore, the database handles parallelism for us already.
+        #
+        # If this raises IntegrityError, it means that this was one of multiple
+        # threads trying to create the same Alert at the same time (the row
+        # not existing at the start of the execution of create_alert() for all
+        # threads, but already existing for one of the threads at the point of
+        # executing save()).
+        #
+        # Therefore, the Alert created by the other thread is very recent.
+        #
+        # Since all Alerts of the same name have practically the same meaning,
+        # it is fine to give up in this thread without retrying to rewrite the
+        # just-created Alert by practically the same information.
+        #
+        # Not using select_for_update() because we don't care if an alert (row) is
+        # overwritten, because all such versions have the same meaning.
+        #
+        # References:
+        # - https://www.postgresql.org/docs/16/index-unique-checks.html#:~:text=At%20present%2C%20only-,b%2Dtree,-supports%20it.)%20Columns
+        # - https://www.postgresql.org/docs/16/locking-indexes.html#:~:text=B%2Dtree%2C,without%20deadlock%20conditions.
+        # - https://www.postgresql.org/files/developer/concurrency.pdf
+
+        try:
+            Alert.objects.create_alert(
+                name=name,
+                object_id=self.uuid,
+                content_type=ContentType.objects.get_for_model(self),
+                description=description,
+                alert_type=alert_type,
+                resolution_steps=resolution_steps,
+                acl_read=acl_read,
+                acl_write=acl_write,
+            ).save()
+        except IntegrityError:
+            pass
```

Comment:
I am not sure if writing general information into code brings any value. The code becomes longer, and the important part is buried under less important parts. If I understood correctly, the only necessary information is why there is `pass` in `except`. And that could be simplified as follows:
```
        except IntegrityError:
            # skip as alerts of the same name have the same meaning
            pass
```

Reply:
OK, changed the comment as you suggested, but kept the change in a separate commit that I'd like _not_ being squashed. The reason is that this was a footgun we spent days on and we're not exactly experts on distributed computing. If there's an issue in it, there's a chance the explanation in git history will help. At least I personally use `git log -S` and other ways of traversing history for pieces of code difficult to understand, and it would help at least me, if I knew this.

Reply:
I think the simplified comment completely throws away the reason why it was done. If I would be someone without context I would have no idea what is going on and why it matters that same name alerts have same meaning.

Reply:
I am okay with having it more detailed in a reasonable way, but a paragraph with 20 lines seems to me to be too much.

Reply:
The original @jobselko 's feedback indicated a slight misunderstanding caused by a low information density of the original block of comments (it would be `alerts of the same name and age have the same meaning` without the slight misunderstanding).

Since both @JakubFrejlach and I prefer to have this footgun documented, I rewrote the block of comments with higher information density.

Now I'm merging it so that I can continue working on OSIDB-3048 (Blocker priority) and test it on stage. Feel free to submit a follow-up PR.